core/src/main/java/hudson/model/Queue.java

```diff
@@ -1478,19 +1493,27 @@ public void maintain() {
                     continue;
                 }
 
+                String taskDisplayName = p.task.getFullDisplayName();
+
                 if (p.task instanceof FlyweightTask) {
                     Runnable r = makeFlyWeightTaskBuildable(new BuildableItem(p));
                     if (r != null) {
                         p.leave(this);
+                        LOGGER.log(Level.FINEST, "Executing flyweight task {0}", taskDisplayName);
                         r.run();
                         updateSnapshot();
                     }
                 } else {
 
                     List<JobOffer> candidates = new ArrayList<JobOffer>(parked.size());
-                    for (JobOffer j : parked.values())
-                        if (j.canTake(p))
+                    for (JobOffer j : parked.values()) {
+                        if (j.canTake(p)) {
+                            LOGGER.log(Level.FINEST,
+                                    "{0} is a potential candidate for task {1}",
+                                    new Object[]{j.executor.getDisplayName(), taskDisplayName});
```

Comment:
I'm still not happy about the displayName calculation without Logging level check clause